File: javascript/packages/linter/src/rules/erb-strict-locals-required.ts
import { ParserRule } from "../types.js"
import { Location, ERBStrictLocalsNode, LiteralNode } from "@herb-tools/core"
import { BaseRuleVisitor } from "./rule-utils.js"

import { isPartialFile } from "./file-utils.js"

import type { UnboundLintOffense, LintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { ParseResult, DocumentNode } from "@herb-tools/core"

class ERBStrictLocalsRequiredVisitor extends BaseRuleVisitor {
  foundStrictLocals: boolean = false

  visitERBStrictLocalsNode(_node: ERBStrictLocalsNode): void {
    this.foundStrictLocals = true
  }
}

export class ERBStrictLocalsRequiredRule extends ParserRule {
  static unsafeAutocorrectable = true
  static ruleName = "erb-strict-locals-required"
  static introducedIn = this.version("0.8.8")

  get parserOptions() {
    return { strict_locals: true }
  }

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: false,
      severity: "error",
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    if (isPartialFile(context?.fileName) !== true) return []

    const visitor = new ERBStrictLocalsRequiredVisitor(this.ruleName, context)
    visitor.visit(result.value)

    if (visitor.foundStrictLocals) return []

    const document = result.value as DocumentNode
    const firstChild = document.children[0]
    const end = firstChild ? firstChild.location.end : Location.zero.end

    return [
      this.createOffense(
        "Partial is missing a strict locals declaration. Add `<%# locals: (...) %>` at the top of the file.",
        Location.from(1, 0, end.line, end.column)
      )
    ]
  }

  autofix(_offense: LintOffense, result: ParseResult): ParseResult | null {
    (result.value.children as unknown[]).unshift(LiteralNode.from({
      type: "AST_LITERAL_NODE",
      location: Location.zero,
      errors: [],
      content: "<%# locals: () %>\n\n",
    }))

    return result
  }
}

File: rust/herb-linter/src/rules/erb_strict_locals_required.rs
use crate::autofix::literal_node;
use crate::offense::{Offense, UnboundOffense};
use crate::rule::{LintContext, ParserRule, Rule};
use crate::utils::file_utils::is_partial_file;

use herb::nodes::{AnyNode, DocumentNode, ERBStrictLocalsNode};
use herb::{Location, ParseResult, Position, Visitor};
use herb_config::{Severity, SeverityConfig};

pub struct ERBStrictLocalsRequiredRule;

#[derive(Default)]
struct StrictLocalsFinder {
  found: bool,
}

impl Visitor for StrictLocalsFinder {
  fn visit_erb_strict_locals_node(&mut self, _node: &ERBStrictLocalsNode) {
    self.found = true;
  }
}

impl Rule for ERBStrictLocalsRequiredRule {
  fn name(&self) -> &'static str {
    "erb-strict-locals-required"
  }

  fn default_severity(&self) -> SeverityConfig {
    SeverityConfig::Severity(Severity::Error)
  }

  fn default_enabled(&self) -> bool {
    false
  }

  fn has_autofix(&self) -> bool {
    true
  }

  fn unsafe_autocorrectable(&self) -> bool {
    true
  }

  fn parser_options(&self) -> herb::ParserOptions {
    herb::ParserOptions {
      strict_locals: true,
      ..crate::rule::default_linter_parser_options()
    }
  }
}

impl ParserRule for ERBStrictLocalsRequiredRule {
  fn check(&self, result: &ParseResult, context: &LintContext) -> Vec<UnboundOffense> {
    if !context.file_name.as_deref().map(is_partial_file).unwrap_or(false) {
      return Vec::new();
    }

    let mut finder = StrictLocalsFinder::default();

    finder.visit_document_node(&result.value);

    if finder.found {
      return Vec::new();
    }

    let end = result
      .value
      .children
      .first()
      .map(|child| child.location().end.clone())
      .unwrap_or_else(|| Position::new(0, 0));

    vec![UnboundOffense::new(
      self.name(),
      "Partial is missing a strict locals declaration. Add `<%# locals: (...) %>` at the top of the file.",
      Location::new(Position::new(1, 0), end),
    )]
  }

  fn autofix(&self, _offense: &Offense, document: &mut DocumentNode, _context: &LintContext) -> bool {
    document
      .children
      .insert(0, AnyNode::LiteralNode(Box::new(literal_node("<%# locals: () %>\n\n"))));

    true
  }
}
